lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -101,6 +101,49 @@ EscapeAnalysis::findRecursivePointerKind(SILType Ty,
   return NoPointer;
 }
 
+// Return the PointerKind that summarized an object's stored properties.
+//
+// If a class only holds fields of non-pointer types, then it is guaranteed not
+// to point to any other objects.
+EscapeAnalysis::PointerKind
+EscapeAnalysis::findObjectKind(SILType Ty, const SILFunction &F) const {
+  if (Ty.isAddress())
+    return AnyPointer;
+
+  auto *classDecl = Ty.getClassOrBoundGenericClass();
+  if (!classDecl)
+    return AnyPointer;
+
+  auto &M = F.getModule();
+  auto expansion = F.getTypeExpansionContext();
+
+  // Start with the most precise pointer kind
+  PointerKind objectKind = NoPointer;
+  auto meetAggregateKind = [&](PointerKind otherKind) {
+    if (otherKind > objectKind)
+      objectKind = otherKind;
+  };
+  for (Type classTy = Ty.getASTType(); classTy;
+       classTy = classDecl->getSuperclass()) {
+    classDecl = classTy->getClassOrBoundGenericClass();
+    assert(classDecl && "superclass must be a class");
+
+    // Return AnyPointer unless we have guaranteed visibility into all class and
+    // superclass properties.
+    if (classDecl->isResilient(M.getSwiftModule(), F.getResilienceExpansion()))
+      return AnyPointer;
+
+    SILType objTy =
+      F.getLoweredType(Lowering::AbstractionPattern::getOpaque(), classTy);
+    for (VarDecl *property : classDecl->getStoredProperties()) {
+      SILType fieldTy =
+          objTy.getFieldType(property, M, expansion).getObjectType();
+      meetAggregateKind(findCachedPointerKind(fieldTy, F));
```

Comment:
That logic looks wrong to me. In case a stored property is a reference, shouldn't it return ReferenceOnly? Instead it recurses into the referenced class type and will e.g. return NoPointer if the referenced class has only trivial properties.

Reply:
I renamed the new helper to `findClassPropertiesPointerKind` with this comment

```
    // For each field in the class, get the pointer kind for that field. For
    // reference-type properties, this will be ReferenceOnly. For aggregates, it
    // will be the meet over all aggregate fields.
```

Reply:
Ah, I misread: it's calling `findCachedPointerKind` and not `findCachedClassPropertiesKind`